### PR TITLE
Add nintendo switch borealis GUI

### PR DESCRIPTION
**Pull request by H0neyBadger**
GUI app for nintendo switch proposal.
it requires https://github.com/thestr4ng3r/chiaki-docker-switch/pull/2 to build the nro (cf: actions/upload-artifact@v1).
I remain available if you have any question or suggestion/review.
Thank you a lot ;-)

You may also compile it for linux (thanks to borealis project)
```
bash scripts/switch/build.sh linux
```
I really hope you will appreciate the result

### CODE REVIEW

**Comment by H0neyBadger** [comment]
I'm adding WIP in title to fix the coding style 
https://gist.github.com/thestr4ng3r/d0b1de10852ca3d119e8c83375211e4b
I will let you know once it's done

**Comment by thestr4ng3r** [comment]
Very cool! I will try it out and review in depth.

File: switch/include/host.h
Comment (thestr4ng3r) [inline switch/include/host.h:116]:
```suggestion
		{
```
also for following

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Awesome, great job!
The GUI is still a bit rough around the edges and there are a few things I would like to change, in particular with how the dependencies are searched for in cmake and make `CHIAKI_SWITCH_ENABLE_LINUX` a bit more generic, but we can easily do that later after merging.

File: switch/src/gui.cpp
Comment (H0neyBadger) [reply switch/src/gui.cpp]: sorry for this bad commit.
I fixed those issues locally and I forgot to add the file in git.

File: switch/src/settings.cpp
Comment (H0neyBadger) [reply switch/src/settings.cpp:478]: hi, 
I copy pasted the config and it seems to work.
what line termination are you using? 
do you have any steps to reproduce the issue ?

I still investigating this problem

Comment (thestr4ng3r) [reply switch/src/settings.cpp:478]: I didn't try this config on the switch directly. It happens when I launch it on Linux with no config file existing, then select "Account ID". Then the GUI hangs when it is normally supposed to show the keyboard (not sure if this is a bug or limitation of borealis). After killing it with SIGINT and restarting, it segfaults.

Comment (H0neyBadger) [reply switch/src/settings.cpp:478]: OK I see,
sorry I did not inform you about this.

from the nintendo switch there is a built-in function that show a keyboad overlay.
the linux version is just for testing and validation. 

for linux you have to use the shell's stdin to push your inputs.

there is a similar issue with sdl controller events.
all are mapped with the switch keys in mind. 
if you try to use a controller on linux all imputes are scrambled

> After killing it with SIGINT and restarting, it segfaults. 
very interesting, I 'm checking

Comment (H0neyBadger) [reply switch/src/settings.cpp:478]: I'm sorry, but I failed to reproduce the issue.
I tried with SIGKILL SIGINT ... with trash in the stdin ... and with trash in chiaki.conf.
the SIGINT close the stdin and write the file locally.
https://github.com/H0neyBadger/chiaki/blob/019d7df856d0c724d221c342722b78c650f79d6b/switch/src/io.cpp#L338
there is no verification for the linux version.  
is this 100% reproducible ?

to be honest, I'm not proud of the settings class.
it's "hand written/homemade" and may lead to many issues.

I m torn between two solution about this config/settings class.
1) use an external lib to manage config file. (but it makes the chiaki project heavier)
2) create a custom class to save our rp key. (our project remains dependency free and easier to port. but on the other hand it contains more bugs) 

let me know if you have any suggestion

**Comment by oushii** [comment]
hello
where to get nor file？

**Comment by H0neyBadger** [comment]
Hello @oushii,
the project is not released yet.

but as a github user you may access to build's artifacts (use it at your own risk)
https://github.com/thestr4ng3r/chiaki/pull/349/checks?check_run_id=1302680146
you may also find the release candidate on https://gbatemp.net/threads/chiaki-ps4-remote-play-homebrew.564553/ first page

for usage and general purposes questions please use the gbatemp thread first.
github is dedicated to bug/enhancement/code.
thank you for your understanding ;-) (I will also answer your question on gbatemp about the keyboard)

File: switch/src/settings.cpp
Comment (thestr4ng3r) [reply switch/src/settings.cpp:478]: Yeah it seems to be 100% reproducible for me. The crash happens only when I re-open the application the second time, not after the signal. `current_host` is null there. But this shouldn't block merging the pr. I can also try to fix it locally since I can reproduce it, just didn't check it in detail yet.

About the current settings implementation, I am not a huge fan of it either. In the Android Version, there is the possibility to import/export settings from/to a json-based format: https://github.com/thestr4ng3r/chiaki/blob/master/android/app/src/main/java/com/metallic/chiaki/common/SerializedSettings.kt
The Qt GUI doesn't have that yet, but it's planned there too. The idea is that you have an easy way of transferring the settings between all the different ports, so the switch version should eventually get that as well. In that sense, we will need a json parser sooner or later anyway, so we could also use it for the settings here, we just should be careful in which json parser library we pick since there are many of them with different properties.
Maybe it would even be possible to make the settings in the same format as the common json so you can just drop a json from your PC on the SD card and have all your registered consoles available.

**Comment by thestr4ng3r** [comment]
Thank you very much!

File: switch/src/settings.cpp
Comment (H0neyBadger) [reply switch/src/settings.cpp:478]: I aggre, portable and standard config is definitely a good design 

currently borealis is including a copy of the https://github.com/nlohmann/json
https://github.com/natinusala/borealis/blob/master/library/include/borealis/extern/nlohmann/json.hpp

I'm ok to follow your advice. just let me know if you have any structure or standard in mind. I will do my best to integrate quickly (but safely).

**Comment by H0neyBadger** [comment]
pleasure is mine. 
I m so glad to work on this project.
thank to you for this project, your help, your support, and for allowing me to contribute on this.
;-) do not hesitate to dispatch issue on it. I will do my best to propose a PR quickly.

**Pull request merged**
